### PR TITLE
Switching the development to virtualenv, pip and tox

### PR DESCRIPTION
**Pull request by henri-hulski**

### CODE REVIEW

File: doc/developing.rst [hunk 1]
```diff
@@ -12,50 +12,41 @@ Install Morepath for development
 
 .. highlight:: sh
 
-First make sure you have virtualenv_ installed for Python 3.5.
+Clone Morepath from github and go to the morepath directory::
 
-.. _virtualenv: https://pypi.python.org/pypi/virtualenv
+  $ git clone git@github.com:morepath/morepath.git
```
Comment (taschini) [inline doc/developing.rst]: If you are using the dollar sign to mark that it is a shell line the highlight on line 13 should be

```
 .. highlight:: console
```

``sh`` is used for shell scripts or things that look like them.

File: doc/developing.rst [hunk 2]
```diff
@@ -12,50 +12,41 @@ Install Morepath for development
...
-Now we need to run bootstrap.py to set up buildout, using the Python from the
-virtualenv we've created before::
+  $ source env/bin/activate
 
-  $ /path/to/ve_morepath/bin/python bootstrap.py
+Install the various dependencies and tools from requirements/develop.txt::
 
-This installs buildout, which can now set up the rest of the development
-environment::
+  $ pip install -r requirements/develop.txt
```
Comment (taschini) [inline doc/developing.rst]: I think it's better here to have

```console
$ pip install -r requirements/develop.txt --src src
```

which installs the editable packages in the ``src`` directory (see the [docs](https://pip.pypa.io/en/stable/reference/pip_wheel/#cmdoption--src)).

Suppose you want to have more than one environments as @faassen said he'd like to have, say py27 and py35.
If you omit the ``--src`` option you'll end up with two different checkouts of Reg, one in ``path/to/py27/src/reg`` and the other in ``path/to/py35/src/reg``.

Comment (henri-hulski) [inline doc/developing.rst]: Thanks! I was thinking about that when set up the tox requirements where it doesn't make sense. But here it's perfect. 
Will add this later as today I'm traveling.

File: doc/developing.rst (hunk 1 above)
Comment (henri-hulski) [inline doc/developing.rst]: Not really part of this PR but sure, can adjust that.

**Comment by faassen** [comment]
Thanks for looking into this, this starts to look like it can replace buildout.

One question I have is about the 'requirements' structure, which is split up quite a bit, but the only thing that I see that uses these bits is tox. What is the reason for the splitting?

**Comment by henri-hulski** [comment]
I think I gonna rework that. 
I want the requirements actually be setup.py. 
I split up the dependencies for tox, so that every testenv install only the ones it needs. But tox can use the extras from setup.py directly. 
The only exception is test.txt as it also installs some more stuff. 
And I've to adjust the reg dependency in

**Comment by henri-hulski** [comment]
Reworked this PR and included propositions from @taschini and @faassen.
It depends on reg >= 0.9.4 so it should be merged after releasing Reg.
Now also removed buildout.

File: doc/developing.rst [hunk 3]
```diff
@@ -10,52 +10,55 @@ on how to get in touch!
...
-Clone Morepath from github and go to the morepath directory::
+  $ source env/py3/bin/activate
 
-  $ git clone git@github.com:morepath/morepath.git
-  $ cd morepath
+Install the various dependencies and development tools from
+develop_requirements.txt::
+
+  $ pip install -r develop_requirements.txt --src src
+
```
Comment (faassen) [inline doc/developing.rst]: Explain briefly what ``src`` is for here.

File: doc/developing.rst [hunk 4]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+together with morepath, they're available in the src directory.
+There you can edit them and test the changes with morepath, which uses them as
+dependencies when you install Morepath for developement as described above.
+
+Installing tox locally with with different Python versions
+----------------------------------------------------------
+
+If you want to check if Morepath works works with the supported Python versions
+and check if it pathes all Travis Ci tests, you can install tox locally.
+
```
Comment (faassen) [inline doc/developing.rst]: pathes -> passes.

I don't understand what you mean by "install tox locally". Oh, I think you mean instead of doing the continuous integration? I think a developer almost always needs to be able to run things with tox locally, so we should just call this "installing tox" and forget about the word "locally".

File: doc/developing.rst [hunk 5]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+Developing Reg, Dectate or Importscan
+-------------------------------------
+
+If you need to adjust the sources of reg, dectate or importscan and test them
+together with morepath, they're available in the src directory.
+There you can edit them and test the changes with morepath, which uses them as
+dependencies when you install Morepath for developement as described above.
+
+Installing tox locally with with different Python versions
```
Comment (faassen) [inline doc/developing.rst]: See comment below, this section should simply be called "tox" I think.

File: doc/developing.rst [hunk 6]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+----------------------------------------------------------
+
+If you want to check if Morepath works works with the supported Python versions
```
Comment (faassen) [inline doc/developing.rst]: "To be able to easily check whether Morepath works with all Python versions it supports"

File: doc/developing.rst [hunk 7]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+and check if it pathes all Travis Ci tests, you can install tox locally.
+
+Tox is also used by Travis Ci on GitHub.
+
+First you should install all Python versions, which you want to test. The
+versions which are not installed will be skipped. At least you should install
+Python 3.5, which is required by flake8, coverage and doctests, and Python 2.7
+for testing Morepath with Python 2.
+
+You can install the Python versions, which are missing in your system with
```
Comment (faassen) [inline doc/developing.rst]: I think the section on pyenv should be removed, as it goes into too much detail on something that's not really essential (I have Pythons installed another way). Instead, say something like "One tool you can use to install multiple versions of Python is pyenv" with a link to its documentation.

File: doc/developing.rst [hunk 8]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+  $ pyenv install 3.5.2
+  $ pyenv install 3.3.6
+  $ pyenv install pypy-5.3.1
+
+And activate them globally, the first one will be the default
+Python environment::
+
+  $ pyenv global 3.5.2 3.3.6 pypy-5.3.1
+
+Make sure you have recent virtualenv, setuptools and pip installed
```
Comment (faassen) [inline doc/developing.rst]: tox doesn't require that these are installed system wide, right? I have tox installed in a virtualenv myself. I think we should explain how to set up a virtualenv with tox here, instead.

File: doc/developing.rst [hunk 9]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+
+.. _pyenv: https://github.com/yyuu/pyenv
+
+.. _Homebrew: https://github.com/yyuu/pyenv#homebrew-on-mac-os-x
+
+.. _`pyenv-installer`: https://github.com/yyuu/pyenv-installer#pyenv-installer
+
+Tox test environments
+---------------------
+
```
Comment (faassen) [inline doc/developing.rst]: Hm, this text is going to be hard to maintain if we change the tox.ini. ``tox -l`` shows a list of test environments, I think you want to mention that. It'd be nicer if tox could somehow document these environments, but I think it's reasonable this documentation into comments in the tox.ini for each section.

File: doc/developing.rst [hunk 10]
```diff
@@ -126,7 +122,114 @@ Building the HTML documentation
...
+You can run all tox tests with::
+
+  $ tox
+
+You can also specify a test environment to run::
+
+  $ tox -e py35
+  $ tox -e pep8
+  $ tox -e docs
```
Comment (faassen) [inline doc/developing.rst]: There are quite a few tox sections now, but I think with my comments (get rid of the pyenv stuff, the tox environment documentation) you can shrink this into a single section that just explains how to install tox (without mention of it being "locally", as that's what this document is about anyway).

**Comment by faassen** [comment]
Okay, I left a bunch of comments concerning the documentation. I think it'll allow you to shrink the tox documentation down a lot, which makes it less intimidating.

**Comment by faassen** [comment]
Do the git+git checkout URLs work if you don't have checkin rights to Morepath? Is a ssh setup with git required? We want to make sure that:

* the core developers can install requirements that they can push changes to.

* any other developer can install requirements at all.

**Comment by faassen** [comment]
Another question: what do I do if I want to run the tests of Reg or Dectate as checked out in ``src``? I think the documentation section on developing these needs to ahve the answer. I think it's just "py.test reg" while you're in that subdir, but just makingsure.

**Comment by faassen** [comment]
Another point. It is my understanding that py.test can be configured in such a way that running ``py.test`` without arguments runs the tests without going into the ``src`` directory and running those tests too. Is it doing that now? Can we adjust dectate and reg to do the same thing if they aren't already?

**Comment by henri-hulski** [comment]
@faassen  Implemented your propositions and also rebased on master.

> Do the git+git checkout URLs work if you don't have checkin rights to Morepath? Is a ssh setup with git required? We want to make sure that:
> 
> * the core developers can install requirements that they can push changes to.
> 
> * any other developer can install requirements at all.

This one is difficult to resolve. Switched back to git+https:// 
But now you need a GitHub password for pushing.
Didn't find a better solution.

**Comment by henri-hulski** [comment]
I have another idea. We can just add the commented out  URLs with the git:git@ scheme with a hint for core developers to switch. Then it's possible to push with your key.

File: develop_requirements.txt [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# sources
+-e https://github.com/morepath/reg.git#egg=reg
```
Comment (faassen) [inline develop_requirements.txt]: I'm curious to see what the plan is here, I saw the thing was a work in progress. The goals are:

* us, the core developers with write access to the Morepath repos, can get writeable checkouts.

* a new developer that does not have write access can get a checkout at all. If they have to upload an ssh key to github to get it, so be it, but they should be able to do it without write access. This needs to be tested if we haven't already..

I can see why having two requirements files is problematic as that would require a duplication of the source URLs, something we want to avoid. If this is impossible without two requirements files we need to discuss our options.

File: doc/developing.rst [hunk 12]
```diff
@@ -108,25 +116,79 @@ Running the documentation tests
...
+
+If you want to run the tests for one of them, let's say Reg, do::
+
+  $ cd src/reg
+  $ py.test reg
+
+Tox
+---
+
+If you want to check if Morepath works works with the supported Python versions
```
Comment (faassen) [inline doc/developing.rst]: There is no "if" about it. "To check that Morepath works" (and only one "works")

File: doc/developing.rst [hunk 13]
```diff
@@ -108,25 +116,79 @@ Running the documentation tests
...
+and check if it passes all Travis Ci tests, you can install tox locally.
```
Comment (faassen) [inline doc/developing.rst]: Remove mention of the Travis Ci stuff here, as you haven't introduced it until the next line. There's no "locally", it's just installing tox.

File: doc/developing.rst [hunk 14]
```diff
@@ -108,25 +116,79 @@ Running the documentation tests
...
+
+Tox is also used by Travis Ci on GitHub.
```
Comment (faassen) [inline doc/developing.rst]: "We have Travis continuous integration installed on Morepath's github repository and it runs the same tox tests after each checkin."

File: doc/developing.rst [hunk 15]
```diff
@@ -108,25 +116,79 @@ Running the documentation tests
...
 
-Building the HTML documentation
--------------------------------
+If you need to adjust the sources of Reg, Dectate or Importscan and
+test them together with Morepath, they're available in the ``src``
+directory. You can edit them and test changes in the Morepath project
+directly.
 
-To build the HTML documentation (output in ``doc/build/html``), run::
+If you want to run the tests for Morepath, Reg, Dectate and Importscan
```
Comment (faassen) [inline doc/developing.rst]: As @taschini pointed out, I actually asked for the opposite. I want you to be able to cd into the Morepath project and do py.test without arguments and have it execute only the Morepath tests. If you then cd into ``src/reg`` and do py.test without arguments, it only runs those tests.

Comment (henri-hulski) [inline doc/developing.rst]: This was the behaviour as it was before.
We had the line in tox.ini:
```ini
[pytest]
testpaths = morepath
```
This is used by py.test.
I will reverse it.

**Comment by henri-hulski** [comment]
@faassen  Implemented your propositions.

Also added commented out URL's for core developers in develop_requirements.txt.

It's not yet ideal, as you have to take care not to push the modified version to Github.
So maybe it's better to have second requirements.txt for core developers. Let's say coredev_requirements.txt.
